### PR TITLE
[wmcb] Add uninstall kubelet command line option

### PR DESCRIPTION
**Pull request by mansikulkarni96**
This PR  adds a command line option "uninstall-kubelet " to uninstall the existing kubelet service from the Windows node. This will help in BYOH implementation to uninstall all services.

### CODE REVIEW

File: internal/test/wmcb/wmcb_test.go [hunk 1]
```diff
@@ -122,6 +124,7 @@ func (vm *wmcbVM) runTest(testCmd string) error {
 		return fmt.Errorf("error running test: %v: %s", err, output)
 	}
 	if strings.Contains(output, "FAIL") {
+		log.Print(output)
```
Comment (mansikulkarni96) [inline internal/test/wmcb/wmcb_test.go:127]: note: remove on next push

File: cmd/bootstrapper/uninstall_kubelet.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
+package main
+
+import (
+	"flag"
+	"os"
+
+	"github.com/openshift/windows-machine-config-bootstrapper/pkg/bootstrapper"
```
Comment (VaishnaviHire) [inline cmd/bootstrapper/uninstall_kubelet.go:7]: this import can be moved to a separate section

File: pkg/bootstrapper/bootstrapper.go [hunk 3]
```diff
@@ -38,6 +38,8 @@ const (
 	KubeletServiceName = "kubelet"
 	// kubeletDependentSvc is the name of the service dependent on kubelet Windows service
 	kubeletDependentSvc = "hybrid-overlay-node"
+	// hybridOverlayDependentSvc is the name of the service dependent on hybrid-overlay-node Windows service
```
Comment (VaishnaviHire) [inline pkg/bootstrapper/bootstrapper.go:41]: Should this change be a part of separate commit ?

**Review by VaishnaviHire — COMMENTED** [review]
Thanks @mansikulkarni96 for working on this. PTAL at my comments

File: cmd/bootstrapper/uninstall_kubelet.go [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+		Run: runUninstallKubeletCmd,
+	}
+)
+
+func init() {
+	rootCmd.AddCommand(uninstallKubeletCmd)
+}
+
+// runUninstallKubeletCmd uninstalls kubelet service from the Windows node
+func runUninstallKubeletCmd(cmd *cobra.Command, args []string) {
```
Comment (aravindhp) [inline cmd/bootstrapper/uninstall_kubelet.go:28]: I suggest following the existing pattern of:
```go
bootstrapper.NewWinNodeBootstrapper()
wmcb.UninstallKubelet()
wmcb.Disconnect()
```
Otherwise you are leaking implementation details here.

File: pkg/bootstrapper/bootstrapper.go [hunk 5]
```diff
@@ -179,20 +181,33 @@ func NewWinNodeBootstrapper(k8sInstallDir, ignitionFile, kubeletPath string, cni
...
-			return nil, fmt.Errorf("error updating kubelet dependents field %v", err)
-		}
-		bootstrapper.kubeletSVC, err = newKubeletService(ksvc, dependents)
-		if err != nil {
-			return nil, fmt.Errorf("could not initialize struct kubeletService: %v", err)
-		}
+	// If there is already a kubelet service running, find and assign it
+	bootstrapper.kubeletSVC, err = AssignExistingKubelet(svcMgr)
+	if err != nil {
+		return nil, fmt.Errorf("could not assign existing kubelet service: %v", err)
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:187]: This refactor can be a separate commit.

File: pkg/bootstrapper/bootstrapper.go [hunk 6]
```diff
@@ -179,20 +181,33 @@ func NewWinNodeBootstrapper(k8sInstallDir, ignitionFile, kubeletPath string, cni
...
 	}
 	return &bootstrapper, nil
 }
 
+// AssignExistingKubelet finds the existing kubelet service from the Windows Service Manager,
+// assigns its value to the kubeletService struct and returns it.
+func AssignExistingKubelet(svcMgr *mgr.Mgr) (*kubeletService, error) {
+	ksvc, err := svcMgr.OpenService(KubeletServiceName)
+	if err != nil {
+		// Do not return error if the service is not already installed
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:197]: This is dangerous as the caller could try to dereference the kubelet service object which is nil at this point. You should rather return a specific error.

File: pkg/bootstrapper/bootstrapper.go [hunk 7]
```diff
@@ -876,17 +891,23 @@ func (cni *cniOptions) configure(kubeletCmd *string) error {
...
-	if err != nil {
-		// Do not return error if the services are not installed.
-		if !strings.Contains(err.Error(), "service does not exist") {
-			return nil, fmt.Errorf("error getting dependent services for kubelet %v", err)
+	// existingDependents is a list of dependent services that exist on the node
+	var existingDependents []*mgr.Service
+	// expectedDependents is a list of dependent services that could exist
+	// on the node, loop through this list to find the existing list of dependents
+	expectedDependents := []string{hybridOverlayDependentSvc, kubeletDependentSvc}
+	// find the existing service dependents
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:899]: For one, this sort of refactor should be a separate commit. It is also not clear why you are considering the transitive dependency, `hybridOverlayDependentSvc`, a direct dependent of the kubelet service.

File: pkg/bootstrapper/kubelet.go [hunk 8]
```diff
@@ -183,6 +183,36 @@ func (k *kubeletService) setRecoveryActions() error {
 	return nil
 }
 
+// remove deletes the kubelet service via the Windows service API
+func (k *kubeletService) remove() error {
```
Comment (aravindhp) [inline pkg/bootstrapper/kubelet.go:187]: Why do you need this method if it is never being used?

File: test/e2e/configure_cni_test.go [hunk 9]
```diff
@@ -25,6 +26,10 @@ func TestConfigureCNI(t *testing.T) {
 	t.Run("Configure CNI", testConfigureCNI)
 }
 
+func TestKubeletUninstall(t *testing.T) {
```
Comment (aravindhp) [inline test/e2e/configure_cni_test.go:29]: It is odd that `TestKubeletUninstall` is added to `configure_cni_test.go`. Shouldn't this be a separate test file?

File: test/e2e/configure_cni_test.go [hunk 10]
```diff
@@ -86,6 +91,22 @@ func testConfigureCNI(t *testing.T) {
...
+		svcMgr, err := mgr.Connect()
+		require.NoError(t, err)
+		defer svcMgr.Disconnect()
+		kubeletSvc, err := bootstrapper.AssignExistingKubelet(svcMgr)
+		require.NoError(t, err)
+		require.NotNil(t, kubeletSvc)
+
+		err = kubeletSvc.UninstallKubeletService()
+		assert.NoError(t, err, "unable to uninstall kubelet service")
+		assert.False(t, svcExists(t, bootstrapper.KubeletServiceName))
```
Comment (aravindhp) [inline test/e2e/configure_cni_test.go:106]: Here too you should be using the pattern:
```go
bootstrapper.NewWinNodeBootstrapper()
wmcb.UninstallKubelet()
wmcb.Disconnect()
```
This is how someone using WMCB as a library would use it.

**Review by aravindhp — CHANGES_REQUESTED** [review]
@mansikulkarni96 thanks for working on this. Please address my comments.

File: cmd/bootstrapper/uninstall_kubelet.go (hunk 4 above)
Comment (mansikulkarni96) [reply cmd/bootstrapper/uninstall_kubelet.go:28]: I agree that was my initial thought but there would be a lot of parameters in calling bootstrapper.NewWinNodeBootstrapper() which would be not be required for uninstall eg: k8sInstallDir, ignitionFile, kubeletPath and would have to be passed as empty strings.

File: pkg/bootstrapper/bootstrapper.go (hunk 6 above)
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:197]: I ll check for a specific error here something like 'not installed'.

File: pkg/bootstrapper/bootstrapper.go (hunk 7 above)
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:899]: Doing this in simplest form possible here, I need to stop kube-proxy-> hybrid-overlay -> kubelet to stop kubelet service due to the dependency order. I can add comments to clarify this and add a seperate commit.

File: pkg/bootstrapper/kubelet.go (hunk 8 above)
Comment (mansikulkarni96) [reply pkg/bootstrapper/kubelet.go:187]: this method is being used in k.stopAndRemove()

**Review by mansikulkarni96 — COMMENTED** [review]
Thank you for the review @aravindhp , I have addressed the comments that could require discussion, will add changes for the rest.

File: cmd/bootstrapper/uninstall_kubelet.go (hunk 4 above)
Comment (aravindhp) [reply cmd/bootstrapper/uninstall_kubelet.go:28]: Think of WMCB as a library and how a user would call. This makes it way too complicated from that perspective. Having empty strings is not a reason for doing this. We can document that the inputs will be ignored if using the uninstall kubelet functionality.

File: pkg/bootstrapper/bootstrapper.go (hunk 7 above)
Comment (aravindhp) [reply pkg/bootstrapper/bootstrapper.go:899]: I feel that the WMCB should return an error and not try to figure out the dependency graph. It should only be aware of the kubelet service. If it cannot stop that then it should return an error. It is up to the caller (WMCO)  to ensure that kube-proxy and hybrid-overlay are stopped before making this call.

File: pkg/bootstrapper/kubelet.go (hunk 8 above)
Comment (aravindhp) [reply pkg/bootstrapper/kubelet.go:187]: I should have been more clear. It is only being used in `k.stopAndRemove()`. Given how little code there is it is fine to just have `k.stopAndRemove()`. But if `k.remove()` was becoming large then yes, you should split it out, which is not the case here.

File: pkg/bootstrapper/kubelet.go [hunk 11]
```diff
@@ -183,6 +183,36 @@ func (k *kubeletService) setRecoveryActions() error {
...
+	}
+	return k.obj.Delete()
+}
+
+// stopAndRemove stops and removes the kubelet service
+func (k *kubeletService) stopAndRemove() error {
+	if k.obj == nil {
+		return nil
+	}
+	k.stop()
```
Comment (aravindhp) [inline pkg/bootstrapper/kubelet.go:199]: Don't you need to check for error here?

File: pkg/bootstrapper/bootstrapper.go (hunk 7 above)
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go:899]: @aravindhp that makes sense, I ll drop handling of dependencies from here.

File: pkg/bootstrapper/kubelet.go (hunk 8 above)
Comment (mansikulkarni96) [reply pkg/bootstrapper/kubelet.go:187]: thanks for clarifying, I ll drop this method.

File: cmd/bootstrapper/uninstall_kubelet.go (hunk 4 above)
Comment (mansikulkarni96) [reply cmd/bootstrapper/uninstall_kubelet.go:28]: that works.

File: pkg/bootstrapper/bootstrapper.go [hunk 12]
```diff
@@ -179,20 +179,35 @@ func NewWinNodeBootstrapper(k8sInstallDir, ignitionFile, kubeletPath string, cni
...
 	}
 	return &bootstrapper, nil
 }
 
+// assignExistingKubelet finds the existing kubelet service from the Windows Service Manager,
+// assigns its value to the kubeletService struct and returns it.
+func assignExistingKubelet(svcMgr *mgr.Mgr) (*kubeletService, error) {
+	ksvc, err := svcMgr.OpenService(KubeletServiceName)
+	if err != nil {
+		// Do not return error if the services are not installed.
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:195]: `service is not installed`

File: cmd/bootstrapper/uninstall_kubelet.go [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+	flag.Parse()
+	// The inputs to NewWinNodeBootstrapper() are ignored while using the uninstall kubelet functionality
```
Comment (aravindhp) [inline cmd/bootstrapper/uninstall_kubelet.go:29]: This doc needs to be added to `NewWinNodeBootstrapper`

File: cmd/bootstrapper/uninstall_kubelet.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+	wmcb, err := bootstrapper.NewWinNodeBootstrapper("", "", "", "",
+		"")
```
Comment (aravindhp) [inline cmd/bootstrapper/uninstall_kubelet.go:31]: Please move to the previous line

File: pkg/bootstrapper/bootstrapper.go [hunk 15]
```diff
@@ -666,6 +666,16 @@ func (wmcb *winNodeBootstrapper) Disconnect() error {
 	return err
 }
 
+// UninstallKubelet uninstalls the kubelet service from Windows node
+func (wmcb *winNodeBootstrapper) UninstallKubelet() error {
+	// Stop and remove kubelet service if it is in Running state.
+	err := wmcb.kubeletSVC.stopAndRemove()
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:672]: Please add a `nil` check for `kubeletSVC`?

File: internal/test/wmcb/wmcb_test.go [hunk 16]
```diff
@@ -109,6 +109,8 @@ func (vm *wmcbVM) runE2ETestSuite(t *testing.T) {
 	vm.runTestBootstrapper(t)
```
Comment (aravindhp) [inline internal/test/wmcb/wmcb_test.go:109]: Shouldn't we add a test that tries to uninstall the kubelet and asserts that it is handled gracefully when no kubelet service is present?

**Review by aravindhp — CHANGES_REQUESTED** [review]
Looks much better now, @mansikulkarni96

File: pkg/bootstrapper/bootstrapper.go [hunk 17]
```diff
@@ -179,7 +179,7 @@ func NewWinNodeBootstrapper(k8sInstallDir, ignitionFile, kubeletPath string, cni
 		}
 	}
 
-	// If there is already a kubelet service running, find and assign it
+	// If there is already a kubelet service running, find it
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:182]: Nit: this should be in the previous commit.

File: pkg/bootstrapper/bootstrapper.go [hunk 18]
```diff
@@ -667,6 +667,19 @@ func (wmcb *winNodeBootstrapper) Disconnect() error {
 	return err
 }
 
+// UninstallKubelet uninstalls the kubelet service from Windows node
+func (wmcb *winNodeBootstrapper) UninstallKubelet() error {
+	if wmcb.kubeletSVC == nil {
+		return fmt.Errorf("kubelet service is not present")
```
Comment (aravindhp) [inline pkg/bootstrapper/bootstrapper.go:673]: We should remember to ignore this error when using this in WMCO.

**Review by aravindhp — APPROVED** [review]
/approve

File: pkg/bootstrapper/bootstrapper.go (hunk 17 above)
Comment (mansikulkarni96) [reply pkg/bootstrapper/bootstrapper.go]: done

File: cmd/bootstrapper/uninstall_kubelet.go [hunk 19]
```diff
@@ -0,0 +1,49 @@
...
+
+	"github.com/openshift/windows-machine-config-bootstrapper/pkg/bootstrapper"
+)
+
+var (
+	// uninstallKubeletCmd describes the uninstall-kubelet command
+	uninstallKubeletCmd = &cobra.Command{
+		Use:   "uninstall-kubelet",
+		Short: "Uninstalls the kubelet service from the Windows node",
+		Long:  "Uninstalls the kubelet service from the Windows node",
```
Comment (sebsoto) [inline cmd/bootstrapper/uninstall_kubelet.go:17]: `Stops and removes the kubelet service` maybe?

Windows documentation seems to use the term `remove` when it comes to services and not `uninstall`.
Dont think mentioning the Windows node is necessary as this is run directly on the Windows VM.

File: cmd/bootstrapper/uninstall_kubelet.go [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+
+	// uninstall kubelet Windows service
+	err = wmcb.UninstallKubelet()
+	if err != nil {
+		log.Error(err, "could not uninstall kubelet service")
+		os.Exit(1)
+	}
+
+	// Send success message to StdOut to ascertain that kubelet removal was successful
+	os.Stdout.WriteString("kubelet service uninstalled successfully")
```
Comment (sebsoto) [inline cmd/bootstrapper/uninstall_kubelet.go:43]: I think its fine to either say `kubelet service removed sucessfully` or `kubelet uninstalled sucessfully`

File: cmd/bootstrapper/uninstall_kubelet.go [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+
+	err = wmcb.Disconnect()
+	if err != nil {
```
Comment (sebsoto) [inline cmd/bootstrapper/uninstall_kubelet.go:46]: nit: I would prefer using this format if  nothing but err is being returned
```suggestion
	if err = wmcb.Disconnect(); err != nil {
```

Comment (sebsoto) [reply cmd/bootstrapper/uninstall_kubelet.go:46]: Applies to https://github.com/openshift/windows-machine-config-bootstrapper/pull/259/files#diff-7bd684eea45943c86e1fa622be0ac5b58ebb684b7059e40d5b332b942b5ed64aR36 as well
`err = wmcb.UninstallKubelet()`

File: test/e2e/uninstall_kubelet_test.go [hunk 22]
```diff
@@ -0,0 +1,46 @@
...
+		}
+
+		wmcb, err := bootstrapper.NewWinNodeBootstrapper("", "", "", "", "")
+		require.NoError(t, err, "could not create wmcb")
+
+		err = wmcb.UninstallKubelet()
+		assert.Error(t, err, "no error when attempting to uninstall kubelet without kubelet svc")
+		assert.Contains(t, err.Error(), "kubelet service is not present", "incorrect error thrown")
+	})
+}
```
Comment (sebsoto) [inline test/e2e/uninstall_kubelet_test.go:46]: You shouldnt be starting test functions with a `t.Run`. In general `t.Run` should be used to divide separate test cases under a common test parent.
In this code you are introducing essentially 2 tests. `Uninstall kubelet without kubelet service present` and `Uninstall Kubelet`. But these tests are being introduced as `TestBootstrapper/Uninstall_kubelet_without_kubelet_service_present/Kubelet_service_uninstalls` and  `TestKubeletUninstall/Uninstall_Kubelet/Kubelet_service_unistalls`. Theres no need for this complexity.

Instead you can have the two new tests be `TestBootstrapper/Kubelet_service_uninstalls` and `TestKubeletUninstallWithoutKubeletSVC`

**Review by sebsoto — CHANGES_REQUESTED** [review]

File: test/e2e/uninstall_kubelet_test.go (hunk 22 above)
Comment (sebsoto) [reply test/e2e/uninstall_kubelet_test.go:46]: `You shouldnt be starting test functions with a t.Run.` This statement is true only in this context, where there is only one test occuring within the the function.

Comment (mansikulkarni96) [reply test/e2e/uninstall_kubelet_test.go:46]: thanks for pointing this out, modified the test calls.

File: test/e2e/uninstall_kubelet_test.go [hunk 23]
```diff
@@ -0,0 +1,42 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/openshift/windows-machine-config-bootstrapper/pkg/bootstrapper"
+)
+
+func TestKubeletUninstall(t *testing.T) {
+	t.Run("Uninstall Kubelet", testKubeletUninstall)
+}
```
Comment (sebsoto) [inline test/e2e/uninstall_kubelet_test.go:14]: I think you can take this one step further and remove this function and rename `testUninstallWithoutKubeletSvc` to `TestUninstallWithoutKubeletSvc`

Comment (mansikulkarni96) [reply test/e2e/uninstall_kubelet_test.go]: I think you meant TestKubeletUninstall, I have already modified TestBootstrapper/testUninstallWithoutKubeletSvc

**Review by sebsoto — COMMENTED** [review]
/lgtm
